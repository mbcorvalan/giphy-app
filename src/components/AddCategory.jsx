import { useState } from 'react'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSearch } from '@fortawesome/free-solid-svg-icons'

export const AddCategory = ({ onNewCategory }) => {
	const [inputValue, setInputValue] = useState('')
	const onInputChange = ({ target }) => {
		setInputValue(target.value)
	}

	const onSubmit = (e) => {
		e.preventDefault()
		const trimmedInputValue = inputValue.trim().replace(/ /g, '+')
		if (trimmedInputValue.length <= 1) return
		onNewCategory(trimmedInputValue)
		setInputValue('')
	}

	return (
		<form className='form-gifts' onSubmit={onSubmit} aria-label='form'>
			<input
				type='text'
				placeholder='Search GIPHY'
				onChange={onInputChange}
				value={inputValue}
				className='form-gifts__input'
			/>
			<div className='form-gifts__icon'>
				<FontAwesomeIcon className='fa-2x' icon={faSearch} />
			</div>
		</form>
	)
}

AddCategory.propTypes = {
	onNewCategory: PropTypes.func.isRequired,
}

AddCategory.defaultProps = {
	onNewCategory: () => {},
}
