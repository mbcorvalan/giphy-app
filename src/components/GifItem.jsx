import PropTypes from 'prop-types'

export const GifItem = ({ title, url }) => {
	return (
		<div className='card-grid__card'>
			<img src={url} alt={title} />
		</div>
	)
}

GifItem.propTypes = {
	title: PropTypes.string.isRequired,
	url: PropTypes.string.isRequired,
}

GifItem.defaultProps = {
	title: '',
	url: '',
}
