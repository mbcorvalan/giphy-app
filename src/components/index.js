export * from './AddCategory'
export * from './GiftGrid'
