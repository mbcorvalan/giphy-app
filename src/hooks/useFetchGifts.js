import { useEffect, useState } from 'react'
import { getGifts } from '../helpers/getGifts'

export const useFetchGifts = (category, page) => {
	const [images, setImages] = useState([])
	const [isLoading, setIsLoading] = useState(true)

	const getImages = async () => {
		const images = await getGifts(category, page)
		setImages(images)
		setIsLoading(false)
	}

	useEffect(() => {
		getImages()
	}, [category, page])

	return {
		images,
		isLoading,
	}
}
