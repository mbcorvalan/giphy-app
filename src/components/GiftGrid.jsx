import { useState } from 'react'
import { GifItem } from './GifItem'
import { useFetchGifts } from '../hooks/useFetchGifts'
import Backdrop from '@mui/material/Backdrop'
import CircularProgress from '@mui/material/CircularProgress'
import Button from '@mui/material/Button'
import ButtonGroup from '@mui/material/ButtonGroup'
import PropTypes from 'prop-types'

export const GiftGrid = ({ category }) => {
	const [page, setPage] = useState(1)
	const { images, isLoading } = useFetchGifts(category, page)

	const nextPage = () => {
		setPage(page + 1)
	}
	const prevPage = () => {
		if (page === 1) return
		setPage(page - 1)
	}

	return (
		<>
			<h3>{category.replace(/\+/g, ' ')}</h3>
			{isLoading && (
				<Backdrop
					sx={{ color: '#fff', zIndex: (theme) => theme.zIndex.drawer + 1 }}
					open={open}
				>
					<CircularProgress color='inherit' />
				</Backdrop>
			)}

			<div className='card-grid'>
				<div className='card-grid__images'>
					{images.map((image) => (
						<GifItem key={image.id} {...image} />
					))}
				</div>
				<div className='card-grid__buttons'>
					<ButtonGroup
						variant='contained'
						aria-label='outlined primary button group'
					>
						<Button disabled={page === 1} size='large' onClick={prevPage}>
							Prev
						</Button>
						<Button size='large' onClick={nextPage}>
							Next
						</Button>
					</ButtonGroup>
				</div>
			</div>
		</>
	)
}

GiftGrid.propTypes = {
	category: PropTypes.string.isRequired,
}

GiftGrid.defaultProps = {
	category: '',
}
