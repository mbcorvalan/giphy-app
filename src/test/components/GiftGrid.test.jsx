import { GiftGrid } from '../../components/GiftGrid'
import { render, screen } from '@testing-library/react'
import { useFetchGifts } from '../../hooks/useFetchGifts'

jest.mock('../../hooks/useFetchGifts')

describe('Test in <GiftGrid />', () => {
	const category = 'One Punch'
	test('should render the component correctly', () => {
		useFetchGifts.mockReturnValue({
			images: [],
			loading: true,
		})
		const wrapper = render(<GiftGrid category={category} />)
		expect(wrapper).toMatchSnapshot()
		/* 	expect(screen.getByText('progressbar')).toBeInTheDocument() */
		expect(screen.getByText(category))
	})
	test('should show items when images are loaded with useFetchGifs', async () => {
		const gifs = [
			{
				id: 'ABC',
				url: 'https://localhost/cualquier/cosa.jpg',
				title: 'Cualquier cosa',
			},
			{
				id: '123',
				url: 'https://localhost/cualquier/cosa.jpg',
				title: 'Cualquier cosa',
			},
		]
		useFetchGifts.mockReturnValue({
			images: gifs,
			loading: false,
		})
		render(<GiftGrid category={category} />)
		expect(screen.getAllByRole('img').length).toBe(gifs.length)
	})
})
