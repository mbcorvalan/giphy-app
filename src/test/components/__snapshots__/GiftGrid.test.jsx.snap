// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <GiftGrid /> should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3>
        One Punch
      </h3>
      <div
        class="card-grid"
      >
        <div
          class="card-grid__images"
        />
        <div
          class="card-grid__buttons"
        >
          <div
            aria-label="outlined primary button group"
            class="MuiButtonGroup-root MuiButtonGroup-contained css-1d9bq45-MuiButtonGroup-root"
            role="group"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge Mui-disabled MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButtonGroup-firstButton css-jh47zj-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              Prev
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButtonGroup-lastButton css-jh47zj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Next
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h3>
      One Punch
    </h3>
    <div
      class="card-grid"
    >
      <div
        class="card-grid__images"
      />
      <div
        class="card-grid__buttons"
      >
        <div
          aria-label="outlined primary button group"
          class="MuiButtonGroup-root MuiButtonGroup-contained css-1d9bq45-MuiButtonGroup-root"
          role="group"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge Mui-disabled MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButtonGroup-firstButton css-jh47zj-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Prev
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButtonGroup-lastButton css-jh47zj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Next
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
